File: leetcode/range-sum-query-2d-immutable.cpp
#include <bits/stdc++.h>
using namespace std;
class NumMatrix
{
 private:
    vector<vector<int>> data;

 public:
    NumMatrix(vector<vector<int>>& matrix) : data(matrix)
    {
        auto data_column_size = data.size();
        auto data_row_size = data[0].size();

        for (int i = 0; i < data_column_size; ++i) {
            for (int j = 0; j < data_row_size; ++j) {
                if (i == 0 && j == 0)
                    continue;

                data[i][j] += get_at(i - 1, j) + get_at(i, j - 1) - get_at(i - 1, j - 1);
            }
        }
    }

    int get_at(int row, int column) { return (row < 0 || column < 0) ? 0 : data[row][column]; }

    int sumRegion(int row1, int col1, int row2, int col2)
    {
        return data[row2][col2] + get_at(row1 - 1, col1 - 1) - get_at(row2, col1 - 1) -
               get_at(row1 - 1, col2);
    }
};

File: leetcode/range-sum-query-2d-immutable.py
from typing import List


class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        rows = len(matrix)
        columns = len(matrix[0])

        self.presum = [[0 for _ in range(columns + 1)] for _ in range(rows + 1)]

        for r in range(rows):
            for c in range(columns):
                self.presum[r + 1][c + 1] = (
                    matrix[r][c]
                    + self.presum[r][c + 1]
                    + self.presum[r + 1][c]
                    - self.presum[r][c]
                )

        print(self.presum)

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return (
            self.presum[row2 + 1][col2 + 1]
            - self.presum[row2 + 1][col1]
            - self.presum[row1][col2 + 1]
            + self.presum[row1][col1]
        )


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
